de.metas.shipper.gateway.dhl/src/main/java/de/metas/shipper/gateway/dhl/DhlShipperGatewayClient.java

```diff
@@ -0,0 +1,420 @@
+import org.springframework.ws.transport.http.HttpComponentsMessageSender;
+import org.springframework.xml.transform.StringResult;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBElement;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.time.format.DateTimeFormatter;
+import java.util.Base64;
+import java.util.List;
+
+public class DhlShipperGatewayClient implements ShipperGatewayClient
+{
+	private static final Logger logger = LoggerFactory.getLogger(DhlShipperGatewayClient.class);
+
+	private final Version API_VERSION;
+
+	private final DhlClientConfig config;
+
+	private final WebServiceTemplate webServiceTemplate;
+	private final SoapHeaderWithAuth soapHeaderWithAuth;
+
+	private final de.dhl.webservice.cisbase.ObjectFactory objectFactoryCis;
+	private final de.dhl.webservices.businesscustomershipping._3.ObjectFactory objectFactory;
+
+	@Builder
+	public DhlShipperGatewayClient(@NonNull final DhlClientConfig config)
+	{
+		this.config = config;
+
+		webServiceTemplate = createWebServiceTemplate();
+		soapHeaderWithAuth = new SoapHeaderWithAuth();
+
+		objectFactoryCis = new de.dhl.webservice.cisbase.ObjectFactory();
+		objectFactory = new de.dhl.webservices.businesscustomershipping._3.ObjectFactory();
+
+		// we're using DHL SOAP API V3
+		API_VERSION = objectFactory.createVersion();
+		API_VERSION.setMajorRelease("3");
+		API_VERSION.setMinorRelease("0");
+	}
+
+	@Override
+	public String getShipperGatewayId()
+	{
+		return DhlConstants.SHIPPER_GATEWAY_ID;
+	}
+
+	@Override
+	public DeliveryOrder createDeliveryOrder(final DeliveryOrder draftDeliveryOrder) throws ShipperGatewayException
+	{
+		throw new ShipperGatewayException("DRAFT Delivery Orders shall never be created.");
+	}
+
+	@Override
+	public DeliveryOrder completeDeliveryOrder(final DeliveryOrder deliveryOrder) throws ShipperGatewayException
+	{
+		logger.trace("Creating shipment order request for {}", deliveryOrder);
+		final CreateShipmentOrderRequest dhlRequest = createDHLShipmentOrderRequest(deliveryOrder);
+
+		final CreateShipmentOrderResponse response = (CreateShipmentOrderResponse)doActualRequest(dhlRequest);
+		final DeliveryOrder completedDeliveryOrder = updateDeliveryOrderFromResponse(deliveryOrder, response);
+		logger.trace("Completed deliveryOrder is {}", completedDeliveryOrder);
+
+		return completedDeliveryOrder;
+	}
+
+	@Override
+	public DeliveryOrder voidDeliveryOrder(final DeliveryOrder deliveryOrder) throws ShipperGatewayException
+	{
+		//noinspection ConstantConditions
+		return null;
+	}
+
+	@Override
+	public List<PackageLabels> getPackageLabelsList(final DeliveryOrder deliveryOrder) throws ShipperGatewayException
+	{
+		logger.trace("getPackageLabelsList for {}", deliveryOrder);
+
+		final DhlCustomDeliveryData customDeliveryData = (DhlCustomDeliveryData)deliveryOrder.getCustomDeliveryData();
```

Comment:
pls avoid casting.
Yes, i know sometimes there is no other way around it.
For those cases i would suggest to have, in DhlCustomDeliveryData a static method:
```java
public static DhlCustomDeliveryData cast(CustomDeliveryData customDeliveryData) { return (DhlCustomDeliveryData)customDeliveryData; }
```

reason:
* easy to track
* easy to discover those cases in case of future refactoring

Reply:
interesting suggestion. will implement!